### PR TITLE
KT 13962: Intention to replace Java collection constructor calls with function calls from stdlib (ArrayList() → arrayListOf())

### PR DESCRIPTION
**Pull request by abansagi**
[Kotlin Issue](https://youtrack.jetbrains.com/issue/KT-13962)
[GitLab Issue](https://gitlab.ewi.tudelft.nl/in4315/2018-2019/TI3125TU-swa-20-kotlin/swa-20-kotlin/issues/14)

Example of this change:

```
fun foo() {
    var list: ArrayList<Int> = ArrayList()
}
```

to
```
fun foo() {
    var list: ArrayList<Int> = arrayListOf()
}
```

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.jetbrains.kotlin.psi.KtPsiFactory
+
+class ConvertCollectionConstructorToFunction : SelfTargetingIntention<KtNameReferenceExpression>(
+    KtNameReferenceExpression::class.java, "Convert Java collection call to function call"
+) {
+
+    private val functionMap = hashMapOf(
+        "ArrayList" to "arrayListOf",
+        "HashMap" to "hashMapOf",
+        "HashSet" to "hashSetOf"
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt]: `LinkedHashMap`, `LinkedHashSet`

https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashSet.html

https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html

The rest are interfaces, I think.

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    override fun applyTo(element: KtNameReferenceExpression, editor: Editor?) {
+
+        val name = element.text ?: return
+
+        val functionName = functionMap[name] ?: return
+
+        val arrayListOfCall = KtPsiFactory(element).createIdentifier(functionName)
+
+        element.replace(arrayListOfCall)
+    }
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt:41]: cut down on new lines I think

File: idea/testData/intentions/convertCollectionConstructorToFunction/keepLinkedHashSetCall.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// WITH_RUNTIME
+// IS_APPLICABLE: false
+
+import java.util.LinkedHashSet
+
+fun foo() {
+    var list: LinkedHashSet<Int> = <caret>LinkedHashSet()
```
Comment (devbridie) [inline idea/testData/intentions/convertCollectionConstructorToFunction/keepLinkedHashSetCall.kt]: why is this `IS_APPLICABLE: false`? It seems that a conversion should be applicable here: `LinkedHashSet` to `linkedSetOf()`.

**Review by devbridie — CHANGES_REQUESTED** [review]

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt (hunk 1 above)
Comment (abansagi) [reply idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt]: I couldn't find a function call for them, you also said it was just those 3 when we went over them, but maybe I overlooked them, I'll look again during the weekend.

File: idea/testData/intentions/convertCollectionConstructorToFunction/keepLinkedHashSetCall.kt (hunk 3 above)
Comment (abansagi) [reply idea/testData/intentions/convertCollectionConstructorToFunction/keepLinkedHashSetCall.kt]: See above ^

**Comment by devbridie** [comment]
* Probably needs `IS_APPLICABLE: false` for cases `ArrayList(1)`, `HashMap(5)`,  `HashSet(4)`, `HashSet(listOf())`, etc

**Comment by devbridie** [comment]
* Don't forget the diamonds.

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun isApplicableTo(element: KtNameReferenceExpression, caretOffset: Int): Boolean {
+        val name = element.text ?: return false
+        return functionMap.containsKey(name)
+    }
+
+    override fun applyTo(element: KtNameReferenceExpression, editor: Editor?) {
+        val name = element.text ?: return
+        val functionName = functionMap[name] ?: return
+        val arrayListOfCall = KtPsiFactory(element).createIdentifier(functionName)
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertCollectionConstructorToFunction.kt]: Rename this variable to `convertedCall`?

**Review by jan-gerling — COMMENTED** [review]
You might add an example to the descriptio of your intention.  I had to read the youtrack issue before I could understand your intention properly.

Otherwise, nice job!

**Comment by devbridie** [comment]
* You shouldn't have commited files in `.idea`.

Otherwise looks good to me.

**Comment by abansagi** [comment]
> * You shouldn't have commited files in `.idea`.

Reverted in 8e725d5

**Review by KilianCallebaut — APPROVED** [review]

**Review by devbridie — APPROVED** [review]

**Pull request merged**